### PR TITLE
feat: add drag and drop column rearrangement for table viz

### LINKED ISSUE #19380: Drag and Drop user-reorganizable columns in the Table Viz
I'm finally bringing this over from the old superset-ui repo. I think this one was almost included, and then they did the whole monorepo.

**Is your feature request related to a problem? Please describe.**
Sometimes the dashboard designer doesn't know how the end user will want their columns organized in a table. Why not let the user rearrange the columns themselves if they want?

**Describe the solution you'd like**
react-table has a built-in useColumnOrder hook, and there was a discussion online about how to implement drag-and-drop in a react-table.
https://github.com/TanStack/react-table/issues/256

**Describe alternatives you've considered**
None, this is the easiest approach.

**Additional context**
I'll link a PR shortly

### PR DESCRIPTION
**Pull request by stevetracvc**
Bringing this over from superset-ui (some discussion was on the old PR)
https://github.com/apache-superset/superset-ui/pull/1423

### SUMMARY
Add the ability for the end user to drag and drop columns for a table visualization

Table designer can enable column rearrangement in the config options when building the table. Then, user can drag columns around

These are **volatile changes**, so when the browser is refreshed, the column order goes back to default. Demo is in the chart edit page, which is just for simple demonstration purposes. This is intended for the end user to be allowed to rearrange the columns.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
**OFF**
dnd_off

**ON**
dnd_on

### ADDITIONAL INFORMATION
- [x] Has associated issue: #19380 
- [x] Fixes #19380 
- [ ] Required feature flags:
- [x] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [x] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by stevetracvc** [comment]
@rusackas @junlincc I finally got around to merging in the monorepo, so I'm reviving this PR :)

File: superset-frontend/plugins/plugin-chart-table/src/controlPanel.tsx [hunk 1]
```diff
@@ -445,6 +445,20 @@ const config: ControlPanelConfig = {
...
+        [
+          {
+            name: 'rearrange_columns',
+            config: {
+              type: 'CheckboxControl',
+              label: t('Allow columns to be rearranged'),
+              renderTrigger: true,
+              default: false,
+              description: t(
+                'Allow the end user to drag-and-drop the column headers to rearrange them.',
```
Comment (ktmud) [inline superset-frontend/plugins/plugin-chart-table/src/controlPanel.tsx:457]:
```suggestion
                "Allow end user to drag-and-drop column headers to rearrange them. Note their changes won't persist for the next time they open the chart.",
```

**Review by ktmud — APPROVED** [review]
This looks great! I have only a couple of nits.

File: superset-frontend/plugins/plugin-chart-table/src/DataTable/DataTable.tsx [hunk 2]
```diff
@@ -210,6 +218,33 @@ export default function DataTable<D extends object>({
...
   const shouldRenderFooter = columns.some(x => !!x.Footer);
 
+  let columnBeingDragged = -1;
+
+  const onDragStart = (e: React.DragEvent) => {
+    const el = e.target as HTMLTableCellElement;
+    columnBeingDragged = allColumns.findIndex(
+      col => col.id === el.dataset.columnName,
+    );
+    e.dataTransfer.setData('text/plain', `${columnBeingDragged}`);
```
[the remark this answers is not in the document]
Comment (stevetracvc) [reply superset-frontend/plugins/plugin-chart-table/src/DataTable/DataTable.tsx:228]: From what I remember, it's what allows the preview of the column header as you drag it

File: superset-frontend/packages/superset-ui-demo/storybook/stories/plugins/plugin-chart-table/TableStories.tsx [hunk 3]
```diff
@@ -66,6 +66,7 @@ function loadData(
     alignPn = false,
     showCellBars = true,
     includeSearch = true,
+    rearrangeColumns = false,
```
[the remark this answers is not in the document]
Comment (stevetracvc) [reply superset-frontend/packages/superset-ui-demo/storybook/stories/plugins/plugin-chart-table/TableStories.tsx:69]: OK, I'll go through and rename it everywhere :)

File: superset-frontend/packages/superset-ui-demo/storybook/stories/plugins/plugin-chart-table/TableStories.tsx [hunk 4]
```diff
@@ -86,6 +87,7 @@ function loadData(
       page_length: pageLength,
       show_cell_bars: showCellBars,
       include_search: includeSearch,
+      rearrange_columns: rearrangeColumns,
```
Comment (ktmud) [inline superset-frontend/packages/superset-ui-demo/storybook/stories/plugins/plugin-chart-table/TableStories.tsx:90]: Can you change the variable name in this file as well?

Comment (stevetracvc) [reply superset-frontend/packages/superset-ui-demo/storybook/stories/plugins/plugin-chart-table/TableStories.tsx]: omg. I grepped the plugin folder for it :laughing:

**Comment by stevetracvc** [comment]
@ktmud let me know if there's anything else you think should be changed!

**Comment by stevetracvc** [comment]
@ktmud same with this one. Anything else you think I need to do for this one?

**Comment by ktmud** [comment]
> @ktmud same with this one. Anything else you think I need to do for this one?

This looks good. Thanks for your contribution! And sorry for missing your message earlier.

**Comment by stevetracvc** [comment]
Awesome, no prob, I'm glad you liked it!

**Pull request merged**